### PR TITLE
feat(mdns): change `mdns::Event` to hold `Vec`

### PR DESCRIPTION
**Pull request by drHuangMHT**
## Description

In previous PR #3606 we've made `mdns::Event` `Clone`, but cloning single-use iterators doesn't sound right. Also you have to create an iterator from the actual data returned before putting it into events. So in this PR the iterators are replaced by `Vec`, as it's the type the data originally come from.

Related #3612.

## Notes & open questions

**breaking change**  
~~Should we remove the use of `SmallVec` altogether? When should we roll out this breaking change?~~

Should we remove `smallvec` entirely from the library? The performance impact is not noticeable before a large P2P network is formed.

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by thomaseizinger** [comment]
I think I'd rather not have `smallvec` in our public API.

If we decide to do this, we can ship it with the next set of breaking changes.

Needs a changelog entry.

**Comment by drHuangMHT** [comment]
> I think I'd rather not have `smallvec` in our public API.

Glad to see that coming. Let's do it at once.

File: protocols/mdns/CHANGELOG.md
Comment (thomaseizinger) [inline protocols/mdns/CHANGELOG.md]:
```suggestion
- Change `mdns::Event` to hold `Vec` and remove `DiscoveredAddrsIter` and `ExpiredAddrsIter`.
  See [PR 3621].
```

File: protocols/mdns/src/behaviour.rs
Comment (thomaseizinger) [inline protocols/mdns/src/behaviour.rs:288]: Can we remove the `smallvec` dependency now?

Comment (drHuangMHT) [reply protocols/mdns/src/behaviour.rs:288]: Not yet. The `Behaviour` still uses `SmallVec`. Want to get rid of that too?

Comment (thomaseizinger) [reply protocols/mdns/src/behaviour.rs:288]: > Not yet. The `Behaviour` still uses `SmallVec`. Want to get rid of that too?

No, was just curious.

**Review by mxinden — COMMENTED** [review]
I am in favor of the removal of the `Iterator`, I am in favor of the removal of `SmalVec`. Thanks @drHuangMHT!

**Comment by drHuangMHT** [comment]
Are we releasing 0.52 right away? Ping @thomaseizinger

**Comment by thomaseizinger** [comment]
> Are we releasing 0.52 right away? Ping @thomaseizinger

Not right away but there are some breaking changes on the horizon, most notable #2680 which will enable kademlia client mode. See #3651 if you want to track progress. Once we decide to merge that, this PR can also go in. I want to do a round of patch releases before that though.

Happy to resolve the merge conflicts for you if you don't want to bother :)

**Comment by drHuangMHT** [comment]
> Happy to resolve the merge conflicts for you if you don't want to bother :)

Thanks. But I'll resume my activity. My server refused to boot because of a faulty off-brand USB hub. I'll try to resolve the conflict myself because it's a chance for me to learn more about git CLI. It's a necessity, I can't get away with that.

**Comment by drHuangMHT** [comment]
Oh no I did not actually resolve the conflict but left the marker here. I thought it was an overlay......

File: protocols/mdns/CHANGELOG.md
Comment (thomaseizinger) [inline protocols/mdns/CHANGELOG.md]:
```suggestion
# 0.44.0 - unreleased
```

**Comment by thomaseizinger** [comment]
We've now opened the merge window for breaking changes so this is good to go in once the merge conflicts are resolved.

**Review by thomaseizinger — APPROVED** [review]
Thanks!

**Pull request merged**
